In [1]:
import pandas as pd

In [2]:
data=pd.read_table(r"C:\Users\Fidha\Downloads\spambase.data",sep=",")

In [3]:
data

0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4   0.5   0.6  ...   0.40  \
0     0.21  0.28    0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.000   
1     0.06  0.00    0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.010   
2     0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.000   
3     0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.000   
4     0.00  0.00    0.00  0.0  1.85  0.00  0.00  1.85  0.00  0.00  ...  0.000   
...    ...   ...     ...  ...   ...   ...   ...   ...   ...   ...  ...    ...   
4595  0.31  0.00    0.62  0.0  0.00  0.31  0.00  0.00  0.00  0.00  ...  0.000   
4596  0.00  0.00    0.00  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...  0.000   
4597  0.30  0.00    0.30  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...  0.102   
4598  0.96  0.00    0.00  0.0  0.32  0.00  0.00  0.00  0.00  0.00  ...  0.000   
4599  0.00  0.00    0.65  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...  0.000   

       0.41  0.42  0.778   0.43   0.44  3.756   61   278  1  
0     0.132   0.0  0.372  0.180  0.048  5.114  101  1028  1  
1     0.143   0.0  0.276  0.184  0.010  9.821  485  2259  1  
2     0.137   0.0  0.137  0.000  0.000  3.537   40   191  1  
3     0.135   0.0  0.135  0.000  0.000  3.537   40   191  1  
4     0.223   0.0  0.000  0.000  0.000  3.000   15    54  1  
...     ...   ...    ...    ...    ...    ...  ...   ... ..  
4595  0.232   0.0  0.000  0.000  0.000  1.142    3    88  0  
4596  0.000   0.0  0.353  0.000  0.000  1.555    4    14  0  
4597  0.718   0.0  0.000  0.000  0.000  1.404    6   118  0  
4598  0.057   0.0  0.000  0.000  0.000  1.147    5    78  0  
4599  0.000   0.0  0.125  0.000  0.000  1.250    5    40  0  

[4600 rows x 58 columns]

In [4]:
data.shape

(4600, 58)

In [5]:
data.head()

0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4   0.5   0.6  ...  0.40  \
0  0.21  0.28    0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.00   
1  0.06  0.00    0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.01   
2  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
3  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
4  0.00  0.00    0.00  0.0  1.85  0.00  0.00  1.85  0.00  0.00  ...  0.00   

    0.41  0.42  0.778   0.43   0.44  3.756   61   278  1  
0  0.132   0.0  0.372  0.180  0.048  5.114  101  1028  1  
1  0.143   0.0  0.276  0.184  0.010  9.821  485  2259  1  
2  0.137   0.0  0.137  0.000  0.000  3.537   40   191  1  
3  0.135   0.0  0.135  0.000  0.000  3.537   40   191  1  
4  0.223   0.0  0.000  0.000  0.000  3.000   15    54  1  

[5 rows x 58 columns]

In [6]:
data.tail()

0  0.64  0.64.1  0.1  0.32   0.2  0.3  0.4  0.5  0.6  ...   0.40  \
4595  0.31   0.0    0.62  0.0  0.00  0.31  0.0  0.0  0.0  0.0  ...  0.000   
4596  0.00   0.0    0.00  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.000   
4597  0.30   0.0    0.30  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.102   
4598  0.96   0.0    0.00  0.0  0.32  0.00  0.0  0.0  0.0  0.0  ...  0.000   
4599  0.00   0.0    0.65  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.000   

       0.41  0.42  0.778  0.43  0.44  3.756  61  278  1  
4595  0.232   0.0  0.000   0.0   0.0  1.142   3   88  0  
4596  0.000   0.0  0.353   0.0   0.0  1.555   4   14  0  
4597  0.718   0.0  0.000   0.0   0.0  1.404   6  118  0  
4598  0.057   0.0  0.000   0.0   0.0  1.147   5   78  0  
4599  0.000   0.0  0.125   0.0   0.0  1.250   5   40  0  

[5 rows x 58 columns]

In [7]:
data.info() # for getting info of each column

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4600 entries, 0 to 4599
Data columns (total 58 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       4600 non-null   float64
 1   0.64    4600 non-null   float64
 2   0.64.1  4600 non-null   float64
 3   0.1     4600 non-null   float64
 4   0.32    4600 non-null   float64
 5   0.2     4600 non-null   float64
 6   0.3     4600 non-null   float64
 7   0.4     4600 non-null   float64
 8   0.5     4600 non-null   float64
 9   0.6     4600 non-null   float64
 10  0.7     4600 non-null   float64
 11  0.64.2  4600 non-null   float64
 12  0.8     4600 non-null   float64
 13  0.9     4600 non-null   float64
 14  0.10    4600 non-null   float64
 15  0.32.1  4600 non-null   float64
 16  0.11    4600 non-null   float64
 17  1.29    4600 non-null   float64
 18  1.93    4600 non-null   float64
 19  0.12    4600 non-null   float64
 20  0.96    4600 non-null   float64
 21  0.13    4600 non-null   float64
 22  

In [8]:
data.describe()

0         0.64       0.64.1          0.1         0.32  \
count  4600.000000  4600.000000  4600.000000  4600.000000  4600.000000   
mean      0.104576     0.212922     0.280578     0.065439     0.312222   
std       0.305387     1.290700     0.504170     1.395303     0.672586   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.420000     0.000000     0.382500   
max       4.540000    14.280000     5.100000    42.810000    10.000000   

               0.2          0.3          0.4          0.5          0.6  ...  \
count  4600.000000  4600.000000  4600.000000  4600.000000  4600.000000  ...   
mean      0.095922     0.114233     0.105317     0.090087     0.239465  ...   
std       0.273850     0.391480     0.401112     0.278643     0.644816  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       0.000000     0.000000     0.000000     0.000000     0.160000  ...   
max       5.880000     7.270000    11.110000     5.260000    18.180000  ...   

              0.40         0.41         0.42        0.778         0.43  \
count  4600.000000  4600.000000  4600.000000  4600.000000  4600.000000   
mean      0.038583     0.139061     0.016980     0.268960     0.075827   
std       0.243497     0.270377     0.109406     0.815726     0.245906   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.065000     0.000000     0.000000     0.000000   
75%       0.000000     0.188000     0.000000     0.314250     0.052000   
max       4.385000     9.752000     4.081000    32.478000     6.003000   

              0.44        3.756           61           278            1  
count  4600.000000  4600.000000  4600.000000   4600.000000  4600.000000  
mean      0.044248     5.191827    52.170870    283.290435     0.393913  
std       0.429388    31.732891   194.912453    606.413764     0.488669  
min       0.000000     1.000000     1.000000      1.000000     0.000000  
25%       0.000000     1.588000     6.000000     35.000000     0.000000  
50%       0.000000     2.275500    15.000000     95.000000     0.000000  
75%       0.000000     3.705250    43.000000    265.250000     1.000000  
max      19.829000  1102.500000  9989.000000  15841.000000     1.000000  

[8 rows x 58 columns]

In [9]:
data.isnull().sum() #checking for null values

0         0
0.64      0
0.64.1    0
0.1       0
0.32      0
0.2       0
0.3       0
0.4       0
0.5       0
0.6       0
0.7       0
0.64.2    0
0.8       0
0.9       0
0.10      0
0.32.1    0
0.11      0
1.29      0
1.93      0
0.12      0
0.96      0
0.13      0
0.14      0
0.15      0
0.16      0
0.17      0
0.18      0
0.19      0
0.20      0
0.21      0
0.22      0
0.23      0
0.24      0
0.25      0
0.26      0
0.27      0
0.28      0
0.29      0
0.30      0
0.31      0
0.32.2    0
0.33      0
0.34      0
0.35      0
0.36      0
0.37      0
0.38      0
0.39      0
0.40      0
0.41      0
0.42      0
0.778     0
0.43      0
0.44      0
3.756     0
61        0
278       0
1         0
dtype: int64

In [10]:
data["1"].value_counts() # to get the count of the labeled column

0    2788
1    1812
Name: 1, dtype: int64

There are 1812 spam email and 2788 non-spam in the given dataset

In [11]:
X = data.iloc[:,:-1] # independent variables
y = data["1"] # target column or dependent variable

In [12]:
X

0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4   0.5   0.6  ...  0.39  \
0     0.21  0.28    0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...   0.0   
1     0.06  0.00    0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...   0.0   
2     0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...   0.0   
3     0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...   0.0   
4     0.00  0.00    0.00  0.0  1.85  0.00  0.00  1.85  0.00  0.00  ...   0.0   
...    ...   ...     ...  ...   ...   ...   ...   ...   ...   ...  ...   ...   
4595  0.31  0.00    0.62  0.0  0.00  0.31  0.00  0.00  0.00  0.00  ...   0.0   
4596  0.00  0.00    0.00  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...   0.0   
4597  0.30  0.00    0.30  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...   0.0   
4598  0.96  0.00    0.00  0.0  0.32  0.00  0.00  0.00  0.00  0.00  ...   0.0   
4599  0.00  0.00    0.65  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...   0.0   

       0.40   0.41  0.42  0.778   0.43   0.44  3.756   61   278  
0     0.000  0.132   0.0  0.372  0.180  0.048  5.114  101  1028  
1     0.010  0.143   0.0  0.276  0.184  0.010  9.821  485  2259  
2     0.000  0.137   0.0  0.137  0.000  0.000  3.537   40   191  
3     0.000  0.135   0.0  0.135  0.000  0.000  3.537   40   191  
4     0.000  0.223   0.0  0.000  0.000  0.000  3.000   15    54  
...     ...    ...   ...    ...    ...    ...    ...  ...   ...  
4595  0.000  0.232   0.0  0.000  0.000  0.000  1.142    3    88  
4596  0.000  0.000   0.0  0.353  0.000  0.000  1.555    4    14  
4597  0.102  0.718   0.0  0.000  0.000  0.000  1.404    6   118  
4598  0.000  0.057   0.0  0.000  0.000  0.000  1.147    5    78  
4599  0.000  0.000   0.0  0.125  0.000  0.000  1.250    5    40  

[4600 rows x 57 columns]

In [13]:
y

0       1
1       1
2       1
3       1
4       1
       ..
4595    0
4596    0
4597    0
4598    0
4599    0
Name: 1, Length: 4600, dtype: int64

#### Transforming values of X using standard scaler

In [14]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
sc.fit(X)
X=sc.transform(X)

## Perceptron Algorithm

In [15]:
# Implementing Perceptron Algorithm
from sklearn.linear_model import Perceptron
model = Perceptron()
model.fit(X, y)

Perceptron()

In [16]:
#importing libraries for evaluating the perfomance

from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from numpy import mean
from numpy import std

In [17]:
#crosss validation
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)

#model evaluation
scores = cross_val_score(model, X, y, scoring='accuracy', cv=cv, n_jobs=-1)
print('Mean Accuracy: ',(mean(scores)))

Mean Accuracy:  0.8774637681159417


In [18]:
# Splitting the data set into two , train and test with a ratio (0.8,0.2)

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=1)

#fitting model with train data
model.fit(X_train,y_train)

Perceptron()

In [19]:
#predcting y values with x_test

y_pred = model.predict(X_test)

In [20]:
from sklearn.metrics import accuracy_score #calculating the accuracy 
print('Accuracy: ',accuracy_score(y_test, y_pred))

Accuracy:  0.8967391304347826


In [21]:
#Tune Perceptron Hyperparameters-Learning Rate

from sklearn.model_selection import GridSearchCV

grid = dict()
grid['eta0'] = [0.0001, 0.001, 0.01, 0.1, 1.0]
search = GridSearchCV(model, grid, scoring='accuracy', cv=cv, n_jobs=-1)
results = search.fit(X,y)

results

GridSearchCV(cv=RepeatedStratifiedKFold(n_repeats=3, n_splits=10, random_state=1),
             estimator=Perceptron(), n_jobs=-1,
             param_grid={'eta0': [0.0001, 0.001, 0.01, 0.1, 1.0]},
             scoring='accuracy')

In [22]:
# evaluation
print(results.best_score_)
print(results.best_params_)

0.8960144927536232
{'eta0': 0.0001}


Model performed well with smaller learning rates .0001 , model achieved best accuracy

In [23]:
# Epoch Rate

grid = dict()
grid['max_iter'] = [1, 10, 100, 1000, 10000]
re_model = Perceptron(eta0=0.0001)
search = GridSearchCV(re_model, grid, scoring='accuracy', cv=cv, n_jobs=-1)
results = search.fit(X,y)
print(results.best_score_)
print( results.best_params_)

0.8960144927536232
{'max_iter': 10}


Epoch rate and best score is obtained

In [24]:
#fitting the best model with tuned parameters

best_model=Perceptron(eta0=0.0001,max_iter=10)
best_model.fit(X_train,y_train)


y_pred = best_model.predict(X_test)
print('Accuracy:',accuracy_score(y_test, y_pred))

Accuracy: 0.8869565217391304


In [25]:
#fitting best model

model_tuned = Perceptron(eta0=0.0001,max_iter=10)
model_tuned.fit(X,y)

#model evaluation
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
scores = cross_val_score(model_tuned, X, y, scoring='accuracy', cv=cv, n_jobs=-1)
print(mean(scores))

0.8960144927536232
